source/simulator/core.cc

```diff
@@ -2147,6 +2162,329 @@ namespace aspect
+              const unsigned int block_vel = introspection.block_indices.velocities;
+              const unsigned int block_p = (parameters.include_melt_transport) ?
+                                           introspection.variable("fluid pressure").block_index
+                                           : introspection.block_indices.pressure;
+              Assert(block_vel == 0, ExcNotImplemented());
+              Assert(block_p == 1, ExcNotImplemented());
+              Assert(!parameters.include_melt_transport
+                     || introspection.variable("compaction pressure").block_index == 1, ExcNotImplemented());
+
+              // create a completely distributed vector that will be used for
+              // the scaled and denormalized solution and later used as a
+              // starting guess for the linear solver
+              LinearAlgebra::BlockVector linearized_stokes_initial_guess (introspection.index_sets.stokes_partitioning, mpi_communicator);
+
+              linearized_stokes_initial_guess.block (block_vel) = current_linearization_point.block (block_vel);
+              linearized_stokes_initial_guess.block (block_p) = current_linearization_point.block (block_p);
+
+              if (nonlinear_iteration == 0)
+                {
+                  current_linearization_point.block(introspection.block_indices.velocities) = 0;
+                  linearized_stokes_initial_guess.block (block_vel) = 0;
+
+                  rebuild_stokes_matrix = assemble_newton_stokes_system = true;
+                  rebuild_stokes_preconditioner = false;
+
+
+                  denormalize_pressure (this->last_pressure_normalization_adjustment,
+                                        linearized_stokes_initial_guess,
+                                        current_linearization_point);
+
+                  compute_current_constraints ();
+                  assemblers.reset (new internal::Assembly::AssemblerLists<dim>());
+                  set_assemblers();
+                  assemble_stokes_system();
+
+                  this->last_pressure_normalization_adjustment = normalize_pressure(current_linearization_point);
+
+                  initial_newton_residual_velo = system_rhs.block(introspection.block_indices.velocities).l2_norm();
+                  initial_newton_residual_pres = system_rhs.block(introspection.block_indices.pressure).l2_norm();
+                  initial_newton_residual = std::sqrt(initial_newton_residual_velo * initial_newton_residual_velo + initial_newton_residual_pres * initial_newton_residual_pres);
+
+                  current_linearization_point.block(introspection.block_indices.velocities) = global_backup_linearization_point.block(introspection.block_indices.velocities);
+
+                  parameters.switch_initial_newton_residual = initial_newton_residual;
+                  parameters.newton_residual = initial_newton_residual;
```

Comment:
I will put it into the NewtonHandler. I have reviewed the current dependencies, I might only need one of the three.